ialirt_data_access/io.py

```diff
@@ -0,0 +1,106 @@
+"""The ``io`` module."""
+
+import contextlib
+import json
+import logging
+import urllib.request
+from urllib.error import HTTPError, URLError
+from urllib.parse import urlencode
+
+import ialirt_data_access
+
+logger = logging.getLogger(__name__)
+
+
+class IALIRTDataAccessError(Exception):
+    """Base class for exceptions in this module."""
+
+    pass
+
+
+@contextlib.contextmanager
+def _get_url_response(request: urllib.request.Request):
+    """Get the response from a URL request.
+
+    This is a helper function to make it easier to handle
+    the different types of errors that can occur when
+    opening a URL and write out the response body.
+    """
+    try:
+        # Open the URL and yield the response
+        with urllib.request.urlopen(request) as response:
+            yield response
+
+    except HTTPError as e:
+        message = (
+            f"HTTP Error: {e.code} - {e.reason}\n"
+            f"Server Message: {e.read().decode('utf-8')}"
+        )
+        raise IALIRTDataAccessError(message) from e
+    except URLError as e:
+        message = f"URL Error: {e.reason}"
+        raise IALIRTDataAccessError(message) from e
+
+
+def _validate_query_params(year: str, doy: str, instance: str):
+    """Validate the query parameters for the IALIRT log API.
+
+    Parameters
+    ----------
+    year : str
+        Year, must be a 4-digit string (e.g., '2024').
+    doy : str
+        Day of year, must be a string between '001' and '365'.
+    instance : str
+        Instance number, must be either '1' or '2'.
+
+    Raises
+    ------
+    ValueError
+        If any parameter is invalid.
+    """
+    if not (year.isdigit() and len(year) == 4):
+        raise ValueError("Year must be a 4-digit string (e.g., '2024').")
+    if not (doy.isdigit() and 1 <= int(doy) <= 365):
+        raise ValueError("DOY must be a string between '001' and '365'.")
+    if instance not in {"1", "2"}:
+        raise ValueError("Instance must be '1' or '2'.")
+
+
+def query(*, year: str, doy: str, instance: str) -> list[str]:
+    """Query the logs.
+
+    Parameters
+    ----------
+    year : str
+        Year
+    doy : str
+        Day of year
+    instance : str
+        Instance number
+
+    Returns
+    -------
+    list
+        List of files matching the query
+    """
+    query_params = {
+        "year": year,
+        "doy": doy,
```

Comment:
I'm not a huge fan of DOY because it has all the drawbacks of ISO in the machine side, but it's not easy for humans to read or use. Is there a reason you went with DOY instead of ISO? (i.e. yyyy-mm-dd)

Reply:
Yeah I hear you. I asked OpsSW about this and they said that this is what they wanted since it is how their files are formatted. They will be the only ones using the download/query for the s3 bucket. But I will make the other APIs for the DynamoDB more user-friendly.

Reply:
Okay, that makes sense! I guess if the ops people want to use DOY that's their right, haha